presto-main/src/main/java/com/facebook/presto/sql/planner/SubPlan.java

```diff
@@ -78,10 +82,17 @@ public void sanityCheck()
                 .map(PlanFragment::getId)
                 .collect(toImmutableMultiset());
 
-        Preconditions.checkState(exchangeIds.equals(childrenIds), "Subplan exchange ids don't match child fragment ids (%s vs %s)", exchangeIds, childrenIds);
+        checkState(exchangeIds.equals(childrenIds), "Subplan exchange ids don't match child fragment ids (%s vs %s)", exchangeIds, childrenIds);
+        long tableWriteCount = stream(Traverser.forTree(PlanNode::getSources).depthFirstPreOrder(fragment.getRoot()))
+                .filter(node -> node instanceof TableWriterNode)
+                .count();
+        checkState(tableWriteCount <= 1, "Fragment cannot contain more than one TableWriterNode");
+        if (tableWriteCount == 1 && !forceSingleNode) {
+            checkState(fragment.getRoot() instanceof TableWriterNode, "TableWriterNode must be the root of a fragment");
```

Comment:
nit: For the sake of the `sanityCheck` signature simplicity i would try to do something like `fragment.getRoot() instanceof TableWriterNode || fragment.getRoot() instanceof OutputNode`